File: frontend/src/types/checkpoint.ts
export type Milestone = "plan" | "code" | "debug" | "test" | "done";

export interface Checkpoint {
  id: string;
  project_id: string;
  milestone: Milestone;
  state_data: Record<string, unknown>;
  metadata: Record<string, unknown>;
  version: number;
  created_at: string;
}

export interface CheckpointDiff {
  from_checkpoint_id: string;
  to_checkpoint_id: string;
  diff: {
    added: Array<Record<string, unknown>>;
    removed: Array<Record<string, unknown>>;
    changed: Array<Record<string, unknown>>;
  };
}

File: backend/app/schemas/checkpoint.py
from typing import Any, Dict

from pydantic import BaseModel, Field


class CheckpointCreate(BaseModel):
    project_id: str
    milestone: str = Field(..., pattern="^(plan|code|debug|test|done)$")
    state_data: Dict[str, Any] = Field(default_factory=dict)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class CheckpointOut(CheckpointCreate):
    id: str
    created_at: str
    version: int
